### DISCUSSION TITLE
Hiding Empty Drawers

### DISCUSSION THREAD

**Discussion by CastleSeven** [yasirkula/UnityRuntimeInspector#34] [Q&A] [answered]
Hi! I was wondering if there was a setting to enable somewhere to hide drawers that have had all of their variables hidden by CustomInspectorSettings. 

E.g. the Animator component in the pic below is empty, so there's no real reason for the drawer to continue to exist:

![image](https://user-images.githubusercontent.com/965241/118276612-79b18e80-b48d-11eb-9bbc-00d47da614f2.png)

**Comment by yasirkula** [comment, marked as answer]
At the moment, unfortunately there isn't. The best you can do is, register to RuntimeInspector's ComponentFilter event to hide the component altogether.

**Comment by CastleSeven** [comment]
For anyone that wants to hack something halfway usable together for this, here's what I did. It seems to work, as all of the Components listed in the Inspector seem to be ObjectFields. So in ObjectFields.cs, I created the following function:


```
private void EvaluateVisibility() {
			int count = 0;
			foreach (var throw_away in Inspector.GetExposedVariablesForType(Value.GetType())) {
				count++;
			}

			if (count > 0)
				this.gameObject.SetActive(true);
			else
				this.gameObject.SetActive(false);
		}
```

I then added a call to this new `EvaluateVisibility `function at the end of the overriden `Refresh()` function for ObjectFields. Seems to work, as components where I've completely turned all variables off are now hidden, and components where I've turned off all variables but intentionally exposed 1 or 2 have those shown correctly. Only very light testing done so far so YMMV. 

![image](https://user-images.githubusercontent.com/965241/118302349-b55a5180-b4a9-11eb-9d04-0ffa8d87060d.png)

**Reply by yasirkula** [reply]
For performance reasons, I'd recommend you to adapt your code to run inside this loop which is executed only when inspected GameObject's components change (instead of at every Refresh): https://github.com/yasirkula/UnityRuntimeInspector/blob/b1041184a24134c585522cf62542a967912cdc02/Plugins/RuntimeInspector/Scripts/RuntimeInspector/Fields/GameObjectField.cs#L101-L106

**Reply by CastleSeven** [reply]
Thanks for the advice @yasirkula -

I altered the function ever so slightly to accept a Type as a parameter, moved it to GameObjectField.cs, and opted to only return a bool instead of do anything to gameobjects:

```
		private bool HasExposedChildren( Type type )
		{
			int count = 0;
			foreach (var throw_away in Inspector.GetExposedVariablesForType(type))
			{
				count++;
			}

			if (count > 0)
				return true;
			else
				return false;
		}
```

Then a simple one line of logic to skip a component if it doesn't have any exposed children:

![image](https://user-images.githubusercontent.com/965241/118319673-b480ea00-b4c0-11eb-9e7d-09e8d123dfee.png)

Thanks again!

**Reply by yasirkula** [reply]
You are welcome! I forgot to mention that if the size of _components_ list doesn't match the number of Component drawers created, all Components drawers will be regenerated at each refresh. So we should remove non exposed Components from the _components_ list. The ideal place for it would be the Refresh function. You can move your code there and remove non exposed Components from the list.
